File: java_client/src/main/java/com/cisc468share/net/Framing.java
package com.cisc468share.net;

import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;

public final class Framing {
    public static final int MAX_FRAME_SIZE = 16 * 1024 * 1024;

    private Framing() {}

    public static byte[] encodeFrame(byte[] payload) {
        if (payload.length > MAX_FRAME_SIZE) {
            throw new IllegalArgumentException("Payload exceeds maximum frame size");
        }
        ByteBuffer buffer = ByteBuffer.allocate(4 + payload.length);
        buffer.putInt(payload.length);
        buffer.put(payload);
        return buffer.array();
    }

    public static byte[] decodeFrame(InputStream inputStream) throws IOException {
        byte[] header = readExact(inputStream, 4);
        int length = ByteBuffer.wrap(header).getInt();
        if (length < 0 || length > MAX_FRAME_SIZE) {
            throw new IOException("Invalid frame size");
        }
        return readExact(inputStream, length);
    }

    private static byte[] readExact(InputStream inputStream, int size) throws IOException {
        byte[] data = new byte[size];
        int offset = 0;
        while (offset < size) {
            int read = inputStream.read(data, offset, size - offset);
            if (read == -1) {
                throw new IOException("Unexpected EOF");
            }
            offset += read;
        }
        return data;
    }
}

File: python_client/net/framing.py
import struct

MAX_FRAME_SIZE = 16 * 1024 * 1024


class FramingError(Exception):
    pass


def encode_frame(payload: bytes):
    if len(payload) > MAX_FRAME_SIZE:
        raise FramingError("payload too large")
    return struct.pack(">I", len(payload)) + payload


def _read_exact(stream, size: int):
    data = b""
    while len(data) < size:
        chunk = stream.read(size - len(data))
        if not chunk:
            raise FramingError("unexpected EOF")
        data += chunk
    return data


def decode_frame(stream):
    size_bytes = _read_exact(stream, 4)
    size = struct.unpack(">I", size_bytes)[0]

    if size > MAX_FRAME_SIZE:
        raise FramingError("frame too large")

    return _read_exact(stream, size)